### PR TITLE
ZENKO-4485: ARWWI iam policies ctst tests

### PR DESCRIPTION
**Pull request by KazToozs**
**What does this PR do, and why do we need it?**
Adds ARWWI tests to Zenko CTST suite using cucumber steps that are [common to iam policy CLI tests](https://github.com/scality/Zenko/pull/1809)

### CODE REVIEW

File: tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    I want to use a web identity
+    And to subscribe to events I want to be notified on
+    And to receive notifications on buckets/objects activities I have subscribed to
+
+  @2.6.0
+  @PreMerge
+  @IAM-Policies-ARWWI
+  Scenario Outline: Assume Role with Web Identity
+    Given an existing bucket "<bucketName>" "<withVersioning>" versioning
+    And we "<haveObject>" an existing object "<objectName>" "<withVersionId>" version-Id "<objectVersionId>"
```
Comment (ghost) [inline tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature]: As discussed in Xin's PR, I would suggest not specifying the name of entities we create, unless we really want to test them: the names **must** be unique for each test and managed by the code only. This will let us run CTST with multiple workers to speed up the testing time, if needed.
The only aspect we can alter is the entities existence (e.g., if the bucket is versioned or not)

```suggestion
    And an object that "<exists>" "<with>" version-Id"
```

Where `exists` can be "does not exist" or "exist" , and `with` can be "with" or "without", then the steps definition can be generic and re-usable easily, and we can also be consistent across scenarios

Comment (ghost) [reply tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature]: if you really need to name the entities yourself, you can leverage the use of data tables. It was not done untill now so not tested as well, but this is supported by the framework that powers CTST: CucumberJS. A nice article to have examples : https://vinodmuralidharan.medium.com/cucumber-js-data-table-guide-a2447b74c86e

File: tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    And a <type> type
+    When the user tries to perform "<ifS3Standard>" "<action>" on the bucket
+    Then the user should be able to perform successfully
```
Comment (ghost) [inline tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature]: Not asking for change here, but just an idea to reduce the number of lines here:
Related to the above comment, if we want to test the APIs with all identities, we can use a data table for all types we want, and then, in the code, we can call sequentially the right prepareForType() and then the API call in a loop, and store everything in `this.results`, the benefit is that we can, this way, automatically test any API as all identity types

File: tests/ctst/steps/iam-policies/common.ts [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+});
+
+Then('the user should be able to perform successfully', function () {
+    this.endForType();
+    if (this.saved.ifS3Standard) {
+        assert.strictEqual(this.result?.err?.includes("AccessDenied"), false);
+    } else {
+        assert.strictEqual(this.result?.statusCode, 200);
+    }
+});
```
Comment (ghost) [inline tests/ctst/steps/iam-policies/common.ts]: Can we check instead that we got a valid response? In case the API crashes with a given identity, we won't get an AccessDenied, but not a valid response either...

File: tests/ctst/world/Zenko.ts [hunk 4]
```diff
@@ -147,18 +472,242 @@ export default class Zenko extends World {
...
+        } catch (err: any) {
+            console.log(err.response);
+            return { statusCode: err.response.status, data: err.response.data };
+        }
+    }
+
+    async restoreObjectResponseCode() {
+        this.addCommandParameter({ restoreRequest: "Days=1" });
+        return await S3.restoreObject(this.getCommandParameters());
+    }
```
Comment (ghost) [inline tests/ctst/world/Zenko.ts]: This would go into another utils file IMHO, this is not really related to the Zenko namespace, but more a convenient way to easily call all APIs with the right parameters using the same step definition.

**Review by ghost — COMMENTED** [review]
See related comments in https://github.com/scality/Zenko/pull/1802#pullrequestreview-1299676396, this is a first review during the draft phase to give some feedback

File: tests/ctst/steps/iam-policies/common.ts (hunk 3 above)
Comment (XinLi0207) [reply tests/ctst/steps/iam-policies/common.ts]: we did consider this part, some API calls just can't return a valid response, like restoreObject, like putObjectRetention, the condition required to get a successfully response for these calls  is not easy to provide, as we are not testing the functionalities of these APIs, so we are thinking of specifying the expected error that is returned for each API call. like for restoreObject it's `invalidObjectState`, for putObjectRetention, it's `InvalidRequest, bucket is missing object lock configuration`. we only accept the expected error returned, if other error returns, we just fail the tests

File: solution/deps.yaml [hunk 5]
```diff
@@ -88,7 +88,7 @@ vault:
   sourceRegistry: registry.scality.com/vault
   dashboard: vault-dashboards
   image: vault
-  tag: 8.7.2
+  tag: 8.7.4
```
Comment (ghost) [inline solution/deps.yaml]: I did it already : https://github.com/scality/Zenko/blob/development/2.6/solution/deps.yaml#L92

It should be gone after rebasing

File: tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+      | PutObjectAcl                    | STORAGE_MANAGER       | with           | exists         |
+      | PutObjectAcl                    | STORAGE_MANAGER       | without        | exists         |
+      | PutBucketTagging                | STORAGE_MANAGER       | without        | does not exist |
+      | DeleteBucketTagging             | STORAGE_MANAGER       | without        | does not exist |
+      | PutBucketReplication            | STORAGE_MANAGER       | with           | does not exist |
+      | MetadataSearch                  | STORAGE_ACCOUNT_OWNER | without        | does not exist |
+      | PutObject                       | STORAGE_ACCOUNT_OWNER | without        | does not exist |
+      | PutObjectAcl                    | STORAGE_ACCOUNT_OWNER | without        | does not exist |
+      | GetObject                       | STORAGE_ACCOUNT_OWNER | without        | exists         |
+      | GetObject                       | STORAGE_MANAGER       | with           | exists         |
```
Comment (ghost) [inline tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature]: This line is duplicated; I can already see it at line 24 and 39 (and here, 56)

Might be worth checking if other lines are not affected?

File: tests/ctst/world/Zenko.ts [hunk 7]
```diff
@@ -53,19 +61,19 @@ export default class Zenko extends World {
...
+                AccessKeyId: this.parameters.AccountAccessKey,
                 SecretAccessKey: this.parameters.AccountSecretKey,
                 SessionToken: this.parameters.AccountSessionToken,
-            }
-            this.cliMode.parameters.AssumedSession = CacheHelper.ARWWI[CacheHelper.AccountName];
-            this.cliMode.assumed = true;
+            };
+            CacheHelper.parameters.AccessKey = this.parameters.AccountAccessKey;
+            CacheHelper.parameters.SecretKey = this.parameters.AccountSecretKey;
+            CacheHelper.parameters.SessionToken = this.parameters.AccountSessionToken;
```
Comment (ghost) [inline tests/ctst/world/Zenko.ts]: Is it still used? I believe we removed the need to pass an ARWWI session to CTST but an account instead, as "AccountSessionToken" does not make sense, as accounts do not have any session tokens

File: tests/ctst/world/Zenko.ts [hunk 8]
```diff
@@ -99,6 +107,322 @@ export default class Zenko extends World {
...
+        const savedParameters = JSON.parse(JSON.stringify(this.cliOptions));
+        this.resetGlobalType();
+        switch (type) {
+            case EntityType.ACCOUNT:
+                await this.prepareRootUser();
+                this.saved.type = EntityType.IAM_USER;
+                break;
+            case EntityType.IAM_USER:
+                await this.prepareIamUser();
+                this.saved.type = EntityType.ACCOUNT;
```
Comment (ghost) [inline tests/ctst/world/Zenko.ts]:
```suggestion
                await this.prepareRootUser();
                this.saved.type = EntityType.ACCOUNT;
                break;
            case EntityType.IAM_USER:
                await this.prepareIamUser();
                this.saved.type = EntityType.IAM_USER;
```

File: tests/ctst/world/Zenko.ts [hunk 9]
```diff
@@ -99,6 +107,322 @@ export default class Zenko extends World {
...
+                break;
+            case EntityType.STORAGE_MANAGER:
+                await this.prepareARWWI(
+                    this.parameters.StorageManagerUsername || 'storage_manager',
+                    this.parameters.KeycloakTestPassword || '123',
```
Comment (ghost) [inline tests/ctst/world/Zenko.ts]: I think I already reviewed this file but in another PR, so putting the comments here again:

We can store the default password in a const

File: tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature (hunk 6 above)
Comment (KazToozs) [reply tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature]: This was due to a set of lines we were copy pasting to add versioned/non-versioned cases for the same APIs, I thought I was being careful but I missed this, good catch, thanks

File: tests/ctst/world/Zenko.ts (hunk 7 above)
Comment (KazToozs) [reply tests/ctst/world/Zenko.ts]: Yes this was because the branch wasn't rebased

File: tests/ctst/world/Zenko.ts (hunk 8 above)
Comment (KazToozs) [reply tests/ctst/world/Zenko.ts]: branch wasn't rebased

File: tests/ctst/world/Zenko.ts (hunk 9 above)
Comment (KazToozs) [reply tests/ctst/world/Zenko.ts]: yes branch wasn't rebased

File: tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature (hunk 6 above)
Comment (KazToozs) [reply tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature]: 892d1348d0be9ca19bd3682978ec006da01430a4

File: tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature [hunk 10]
```diff
@@ -0,0 +1,191 @@
+Feature: Assume Role with Web Identity
+    In order to interact with restricted APIs
+    As an Artesca User
+    I want to use a web identity
+    And to subscribe to events I want to be notified on
+    And to receive notifications on buckets/objects activities I have subscribed to
```
Comment (ghost) [inline tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature]: should be adapted for this file

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Review by XinLi0207 — APPROVED** [review]
